packages/library/src/rpc-request-coalescer-gating.ts

```diff
@@ -0,0 +1,22 @@
+interface IHasParams {
+    params: unknown;
+}
+interface IHasMethod {
+    method: string;
+}
+
+export function shouldCoalesceRequest(payload: unknown): payload is IHasMethod & IHasParams {
+    if (payload == null || typeof payload !== 'object' || Array.isArray(payload)) {
```

Comment:
Sorry, this is my silly lack of knowledge of TypeScript. If you're already specifying `payload is IHasMethod & IHasParams`, how could it not be an object or be an array?

Reply:
That `payload is IHasMethod & IHasParams` assertion is actually on the _return_ type. This function I've written is what's called a ‘type guard.’ We're basically saying that if you assert, anywhere in your program, that the return value of this function is `true` then `payload` is `IHasMethod & IHasParams`;

```ts
function isFish(pet: Pet): pet is Fish {
  return pet.swims === true;
}

function addToFishtank(fish: Fish) {
  /* ... */
}

const pet = getPet();
addToFishtank(pet); // ERROR: Expected `Fish`
if (isFish(pet)) {
  addToFishtank(pet); // OK
}
```

Reply:
So tl;dr `payload` enters that function a pile of who knows what, and we run assertions on it to the point where if the function returns `true`, Typescript will trust that `payload` is a `IHasMethod & IHasParams`.